File: Data Structures & Algorithms/subsets/submission-3.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        # sol = []
        # res = []
        # n = len(nums)

        # def backtrack(i):
        #     if i == n:
        #         res.append(sol[:])
        #         return

        #     backtrack(i+1)
        #     sol.append(nums[i])
        #     backtrack(i+1)
        #     sol.pop()

        # backtrack(0)

        # return res


        res = []
        
        def backtrack(i, sol):
            res.append(sol[:])

            for j in range(i, len(nums)):
                # if j > i and nums[j] == nums[j - 1]:
                #     continue

                sol.append(nums[j])
                backtrack(j + 1, sol)
                sol.pop()
            
        backtrack(0, [])
        return res

File: Data Structures & Algorithms/word-break-ii/submission-3.go
func wordBreak(s string, wordDict []string) []string {
    wordSet := make(map[string]bool)
    for _, word := range wordDict {
        wordSet[word] = true
    }

    sol := []string {}
    res := []string {}
    var backtrack func(i int) 

    backtrack = func(i int) {
        if i == len(s) {
            res = append(res, strings.Join(sol, " "))
            return
        }
        for j:=i; j < len(s); j++ {
            w := s[i : j+ 1]

            if wordSet[w] {
                sol = append(sol, w)
                backtrack(j + 1)
                sol = sol[:len(sol)-1]
            }
        }
    }

    backtrack(0)
    return res
}
